### PR TITLE
Add `get_variables` dispatch for `Reaction`

### PR DESCRIPTION
**Pull request by TorkelE**
The `get_variables` function is currently defined for `Equation`s but not `Reaction`s. I've added it. It goes through all fields, checking and adds all symbolic variables to an output vector.

### CODE REVIEW

**Review by isaacsas — CHANGES_REQUESTED** [review]
If we are going to add it we should probably add the in-place version and then just call that for the allocating version. It might also be worth using a `Set` and converting to a vector at the end instead. See https://github.com/JuliaSymbolics/Symbolics.jl/blob/1da13fdc375fccd6cd0a779a53b06be4ec7a54ec/src/utils.jl#L47-L79

**Comment by isaacsas** [comment]
Actually, we already have an inplace version that takes `Set`s, so why not just wrap that and then convert to a vector at the end?

**Comment by TorkelE** [comment]
In practise, I just used what Symbolics define for `Equation`s:
```julia
get_variables(eq::Equation) = unique(vcat(get_variables(eq.lhs), get_variables(eq.rhs)))
```
and I figured the above definition was most faithful to that.

Little bit uncertain exactly what rewrite you want, but if you have a different way to phrase the function content I am happy to update.

File: src/reaction.jl [hunk 1]
```diff
@@ -343,6 +343,42 @@ end
...
+    - Among substrates and products.
+    - Among stoichiometries.
+    - Among potential noise scaling metadata.
+"""
+function ModelingToolkit.get_variables!(set, rx::Reaction)
+    get_variables!(set, rx.rate)
+    append!(set, rx.substrates)
+    append!(set, rx.products)
+    for stoichs in (rx.substoich, rx.prodstoich), stoich in stoichs
+        get_variables!(set, stoich)
```
Comment (isaacsas) [inline src/reaction.jl:375]: Shouldn't this check if `stoich` is a Symbolic? If not there is no reason to call `get_variables!`?

File: src/reaction.jl [hunk 2]
```diff
@@ -343,6 +343,42 @@ end
...
+encountered in:
+    - Rates.
+    - Among substrates and products.
+    - Among stoichiometries.
+    - Among potential noise scaling metadata.
+"""
+function ModelingToolkit.get_variables!(set, rx::Reaction)
+    get_variables!(set, rx.rate)
+    append!(set, rx.substrates)
+    append!(set, rx.products)
```
Comment (isaacsas) [inline src/reaction.jl:373]:
```suggestion
    foreach(sub -> push!(set, sub), rx.substrates)
    foreach(prod -> push!(set, prod), rx.products)
```
`append!` would restrict this function to only work on vectors (i.e. `Sets` wouldn't work).

File: src/reaction.jl [hunk 3]
```diff
@@ -343,6 +343,42 @@ end
...
+    for stoichs in (rx.substoich, rx.prodstoich), stoich in stoichs
+        get_variables!(set, stoich)
+    end
+    if has_noise_scaling(rx)
+        get_variables!(set, get_noise_scaling(rx))
+    end
+    return unique!(set)
```
Comment (isaacsas) [inline src/reaction.jl:380]:
```suggestion
    return (set isa AbstractVector) ? unique!(set) : set
```
If passing a `Set` don't need `unique!` (not sure if even defined for them).

**Review by isaacsas — CHANGES_REQUESTED** [review]
A few minor changes. Once addressed and tests pass feel free to merge.

File: src/reaction.jl (hunk 2 above)
Comment (TorkelE) [reply src/reaction.jl:373]: That make sense, good catch.

File: src/reaction.jl (hunk 3 above)
Comment (TorkelE) [reply src/reaction.jl:380]: Yes, I didn't really think of the difference between sets and vectors, good point

File: src/reaction.jl (hunk 1 above)
Comment (TorkelE) [reply src/reaction.jl:375]: Yeah, that is definitely an improvement.

**Pull request merged**